assets/scripts/segments/buildings.js

```diff
@@ -18,341 +18,316 @@ export const BUILDING_DESTINATION_THUMBNAIL = 2
+ *
+ * @param {string} variant
+ * @param {string} position - either "left" or "right"
+ * @returns {string}
+ */
+function getSpriteId (variant, position) {
+  const building = BUILDINGS[variant]
+  return building.spriteId + (building.sameOnBothSides ? '' : '-' + position)
+}
+
+/**
+ * Calculate building image height. For buildings that do not have multiple floors, this
+ * is just the image's intrinsic height value. For buildings with multiple floors, this
+ * must be calculated from the number of floors and sprite pixel specifications.
+ *
+ * @param {string} variant
+ * @param {string} position - either "left" or "right"
+ * @param {Number} floors
+ */
+export function getBuildingImageHeight (variant, position, floors = 1) {
+  const building = BUILDINGS[variant]
+  let height
 
-  return {
-    tilePositionX: tilePositionX,
-    tilePositionY: tilePositionY,
-    width: width,
-    variantsCount: variantsCount,
-    tileset: tileset,
-    mainFloorHeight: mainFloorHeight,
-    floorHeight: floorHeight,
-    flooredBuilding: flooredBuilding,
-    floorRoofWidth: floorRoofWidth,
-    floorCount: floorCount,
-    realHeight: realHeight,
-    roofHeight: roofHeight,
-    height: height,
-    buildingVariant: buildingVariant
+  if (building.hasFloors) {
+    height = ((building.roofHeight + (building.floorHeight * (floors - 1)) + building.mainFloorHeight) * TILE_SIZE)
+  } else {
+    const id = getSpriteId(variant, position)
+    const svg = images.get(id)
+    height = svg.height / TILESET_POINT_PER_PIXEL
   }
+
+  return height
 }
 
 /**
- * Returns true if the building type editable number of floors
+ * Converts the number of floors to an actual height in feet
  *
- * @param {string} buildingVariant
- * @param {Boolean}
+ * @param {string} variant
+ * @param {string} position - "left" or "right"
+ * @param {Number} floors
+ * @returns {Number} height, in feet
  */
-// TODO change to array
-export function isFlooredBuilding (buildingVariant) {
-  if ((buildingVariant === 'narrow') || (buildingVariant === 'wide') ||
-    (buildingVariant === 'residential')) {
-    return true
+export function calculateRealHeightNumber (variant, position, floors) {
+  const CURB_HEIGHT = 6
+  return (getBuildingImageHeight(variant, position, floors) - CURB_HEIGHT) / TILE_SIZE
+}
+
+export function drawBuilding (
+  ctx, destination, street,
+  left, totalWidth, totalHeight,
+  offsetLeft,
+  multiplier, dpi) {
+  const variant = left ? street.leftBuildingVariant : street.rightBuildingVariant
+  const floors = left ? street.leftBuildingHeight : street.rightBuildingHeight
+  const position = left ? 'left' : 'right'
+
+  const buildingHeight = getBuildingImageHeight(variant, position, floors)
+  const offsetTop = totalHeight - (buildingHeight * multiplier)
+
+  const building = BUILDINGS[variant]
+
+  if (!building.hasFloors) {
+    drawSingleFloorBuilding(ctx, variant, position, totalWidth, offsetLeft, offsetTop, multiplier, dpi)
```

Comment:
Would it be easy/cleaner to have one function to draw a building and handle the single/multi floor logic in it?

Reply:
That's what it used to do, but separating it out made it easier to trace the logic. It might be worthwhile to put it back into one function if (in the future) it can be further condensed